content/testing.md

```diff
@@ -0,0 +1,683 @@
+
+<h3 id="driver-packages">Driver packages</h3>
+
+When you run a `meteor test` command, you must provide a `--driver-package` argument. A test driver is a mini-application that runs in place of your app and runs each of your defined tests, whilst reporting the results in some kind of user interface.
+
+There are two main kinds of test driver packages:
+
+  - web-reporters which are Meteor applications and display a special test reporting web UI that you can view the test results in.
+
+  <img src="images/mocha-test-results.png">
+
+  - console-reporters that run completely on the command-line and are primary used for automated testing like [continuous integration](#ci) (as we'll see, typically PhantomJS is used to drive such tests).
+
+In this article, we'll use the popular [Mocha](https://mochajs.org) test runner alongside the [Chai](http://chaijs.com) assertion library to test our application. In order to write tests in Mocha, we can add the [`avital:mocha`](https://atmospherejs.com/avital/mocha) package to our app.
+
+```bash
+meteor add avital:mocha
+```
+
+This package also doesn't do anything in development or production mode (in fact it declares itself `testOnly` so it is not even included in those modes), but when our app is run in [test mode](#test-modes), it takes over, executing test code on both the client and server, and rendering results to the browser.
+
+Test files themselves (for example a file named `todos-item.test.js` or `routing.app-specs.coffee`) can register themselves to be run by the test driver in the usual way for that testing library. For Mocha, that's by using `describe` and `it`:
+
+```js
+describe('my module', () => {
+  it('does something that should be tested', () => {
+    // This code will be executed by the test driver when the app is started in the correct mode
+  })
+})
+```
+
+<h2 id="test-data">Test data</h2>
+
+When your app is run in test mode, it is initialized with a clean test database.
+
+If you are running a test that relies on using the database, and specifically the content of the database, you'll need to perform some *setup* steps in your test to ensure the database is in the state you expect. There are some tools you can use to do this.
+
+To ensure the database is clean, the [`xolvio:cleaner`](https://atmospherejs.com/xolvio/cleaner) package is useful. You can use it to reset the database in a `beforeEach` block:
+
+```js
+import { resetDatabase } from 'meteor/xolvio:cleaner';
+
+describe('my module', () => {
+  beforeEach(() => {
+    resetDatabase();
+  });
+});
+```
+
+This technique will only work on the server. If you need to reset the database from a client test, you can use a method to do so:
+
+```js
+import { resetDatabase } from 'meteor/xolvio:cleaner';
+
+// NOTE: Before writing a method like this you'll want to double check this file is only going 
+// to be loaded in test mode!!
+Meteor.methods({
+  'test.resetDatabase': () => resetDatabase();
```

Comment:
Methods shouldn't be written using arrow functions because they often depend on `this`. It's not a big deal here, but we might not want people to get into the habit of using arrow functions and methods together.